libyul/optimiser/LabelIDDispenser.cpp

```diff
@@ -0,0 +1,163 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/LabelIDDispenser.h>
+
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/OptimizerUtilities.h>
+
+#include <fmt/compile.h>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/iota.hpp>
+
+using namespace solidity::yul;
+
+namespace
+{
+bool isInvalidLabel(
+	std::string_view const _label,
+	std::set<std::string, std::less<>> const& _reservedLabels,
+	Dialect const& _dialect
+)
+{
+	return isRestrictedIdentifier(_dialect, _label) || _reservedLabels.contains(_label);
+}
+}
+
+LabelIDDispenser::LabelIDDispenser(ASTLabelRegistry const& _labels, std::set<std::string> const& _reservedLabels):
+	m_labels(_labels),
+	m_reservedLabels(_reservedLabels.begin(), _reservedLabels.end()),
+	m_offset(_labels.maxID() + 1)
+{
+	m_reservedLabels += std::set{""};
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::newID(LabelID const _parent)
+{
+	m_idToLabelMapping.push_back(resolveParentLabelID(_parent));
+	return m_idToLabelMapping.size() - 1 + m_offset;
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::newGhost()
+{
+	m_idToLabelMapping.push_back(ASTLabelRegistry::ghostLabelIndex());
+	return m_idToLabelMapping.size() - 1 + m_offset;
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::resolveParentLabelID(LabelID _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size() + m_offset, "ID exceeds bounds.");
+	// bigger than offset means, that the input label id that was spawned by this dispenser
+	if (_id >= m_offset)
+		_id = m_idToLabelMapping[_id - m_offset];
+	yulAssert(
+		_id < m_offset && !m_labels.unused(_id),
+		"We can have at most one level of indirection and the derived-from label cannot be unused."
+	);
+	return _id;
+}
+
+bool LabelIDDispenser::ghost(LabelID const _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size() + m_offset, "ID exceeds bounds.");
+	if (_id >= m_offset)
+		return m_idToLabelMapping[_id - m_offset] == ASTLabelRegistry::ghostLabelIndex();
+
+	return m_labels.ghost(_id);
+}
+
+
+ASTLabelRegistry LabelIDDispenser::generateNewLabels(Block const&, Dialect const& _dialect) const
+{
+	// this can be replaced by the actually used ids in the provided block once the AST uses ids instead of YulString
+	std::set<LabelID> usedIDs = ranges::views::iota(static_cast<size_t>(0), m_idToLabelMapping.size() + m_offset) | ranges::to<std::set<LabelID>>;
+
+	if (usedIDs.empty())
+		return {};
```

Comment:
This condition will never be satisfied, because `ASTLabelRegistry` always contains at least one label (the empty one) and therefore `usedIDs` will always contain `0`.

Reply:
Another consequence of including `0` in `usedIDs` is that you'll try to generate a new label for it (you also mark it as reused), which will overwrite `idToLabelMap[0]`. I think that'll actually trip an assertion in `ASTLabelRegistry` constructor, because it requires that the label at `0` is `""`.

On the other hand the `""` you put in `labels` will go unreferenced (unless there are unused IDs in the original registry).

Reply:
> This condition will never be satisfied, because `ASTLabelRegistry` always contains at least one label (the empty one) and therefore `usedIDs` will always contain `0`.

Right. Although this is a bit temporary and to be replaced with whatever the `NameCollector` yields on an input block in which it may happen to have an empty `usedIDs` set. I have removed the condition for now so it can be re-introduced when properly integrating the block input.